sdks/java/io/neo4j/src/main/java/org/apache/beam/sdk/io/neo4j/Neo4jIO.java

```diff
@@ -0,0 +1,1358 @@
+import org.apache.beam.sdk.transforms.display.DisplayData;
+import org.apache.beam.sdk.transforms.display.HasDisplayData;
+import org.apache.beam.sdk.values.PCollection;
+import org.apache.beam.sdk.values.PDone;
+import org.apache.beam.sdk.values.TypeDescriptor;
+import org.apache.beam.vendor.guava.v26_0_jre.com.google.common.base.Preconditions;
+import org.checkerframework.checker.nullness.qual.Nullable;
+import org.neo4j.driver.AuthTokens;
+import org.neo4j.driver.Config;
+import org.neo4j.driver.Driver;
+import org.neo4j.driver.GraphDatabase;
+import org.neo4j.driver.Logging;
+import org.neo4j.driver.Record;
+import org.neo4j.driver.Result;
+import org.neo4j.driver.Session;
+import org.neo4j.driver.SessionConfig;
+import org.neo4j.driver.TransactionConfig;
+import org.neo4j.driver.TransactionWork;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * This is a Beam IO to read from, and write data to, Neo4j.
+ *
+ * <p>
+ *
+ * <p>
+ *
+ * <h3>Driver configuration</h3>
+ *
+ * <p>To read from or write to Neo4j you have to provide a {@link DriverConfiguration} using<br>
+ * 1. {@link DriverConfiguration#create()} (which must be {@link Serializable});<br>
+ * 2. or {@link DriverConfiguration#create(String, String, String)} (URL, username and password).
+ *
+ * <p>If you have trouble connecting to a Neo4j Aura database please try to disable a few security
+ * algorithms in your JVM. This makes sure that the right one is picked to connect:
+ *
+ * <p>
+ *
+ * <pre>{@code
+ * Security.setProperty(
+ *       "jdk.tls.disabledAlgorithms",
+ *       "SSLv3, RC4, DES, MD5withRSA, DH keySize < 1024, EC keySize < 224, 3DES_EDE_CBC, anon, NULL");
+ * }</pre>
+ *
+ * <p>
+ *
+ * <p>
+ *
+ * <p>To execute this code on GCP Dataflow you can create a class which extends {@link
+ * JvmInitializer} and implement the {@link JvmInitializer#onStartup()} method. You need to annotate
+ * this new class with {@link com.google.auto.service.AutoService}
+ *
+ * <pre>{@code
+ * @AutoService(value = JvmInitializer.class)
+ * }</pre>
+ *
+ * <p>
+ *
+ * <h3>Reading from Neo4j</h3>
+ *
+ * <p>{@link Neo4jIO#read()} source returns a bounded collection of {@code OuptutT} as a {@code
+ * PCollection<OutputT>}. OutputT is the type returned by the provided {@link RowMapper}. It accepts
+ * parameters as input in the form of {@code ParameterT} as a {@code PCollection<ParameterT>}
+ *
+ * <p>The following example reads ages to return the IDs of Person nodes. It runs a Cypher query for
+ * each provided age.
+ *
+ * <p>The {@link ParametersMapper} maps input values to each execution of the Cypher statement.
+ * These parameter values are stored in a {@link Map} (mapping String to Object). The map is cleared
+ * after each execution.
+ *
+ * <p>The {@link RowMapper} converts output Neo4j {@link Record} values to the output of the source.
+ *
+ * <pre>{@code
+ * pipeline.apply( Neo4jIO.<Integer, String>read()
```

Comment:
This looks great, awesome :) Though I'm unfortunately not the person capable of reviewing this in detail.
However I noticed a few minor things.

- Given `read` requires `PCollection<ParameterT>` as input one cannot naively copy the code in this example. Adding the input could help understanding and avoid confusion. E.g.
  ```Java
  pipeline
    .apply(Create.of(40, 50, 60))
    .apply( Neo4jIO.<Integer, String>read()
     ... )
  ```
- IOs often provide `read()` and `readAll()` (e.g. have a look at JdbcIO). `readAll` typically performs a parameterized read for each element of the input `PCollection`. To be consistent I would suggest to also name `Neo4jIO.read()` accordingly as `Neo4jIO.readAll()`

Reply:
Great advice! Thanks @mosche. The sample basics were taken from JdbcIO but it's easy enough to just add an extra line to clarify the example.  I'll patch that up in a commit.